### ISSUE TITLE
why not include eos_token in action_seq, which may make mistakes?

### ISSUE THREAD

**Issue by ZiyiLiubird** [OpenLLMAI/OpenRLHF#181]
https://github.com/OpenLLMAI/OpenRLHF/blob/01c78c13ee3ac761085d9d1ba1cd830aa83ae4df/openrlhf/models/actor.py#L126
https://github.com/OpenLLMAI/OpenRLHF/blob/01c78c13ee3ac761085d9d1ba1cd830aa83ae4df/openrlhf/models/actor.py#L127
https://github.com/OpenLLMAI/OpenRLHF/blob/01c78c13ee3ac761085d9d1ba1cd830aa83ae4df/openrlhf/trainer/ppo_utils/experience_maker.py#L118
https://github.com/OpenLLMAI/OpenRLHF/blob/01c78c13ee3ac761085d9d1ba1cd830aa83ae4df/openrlhf/models/actor.py#L144

For example, the sequence is "Today is a sunny day \</s>", and the prompt is "Today is" and its length is 2. The response is "a sunny day \</s>". So the num_actions is 3 if action_seq = sequences[:, input_len:-1] , which not includes \</s>. The length of dim 1 of log_probs is 5,  and log_probs[:, -num_actions:] index "sunny day \</s>" instead of "a sunny day" or "a sunny day \</s>".

**Comment by ZiyiLiubird** [comment]
> we also removed `</s>` in log_probs, so the log_probs[:, -num_actions] should be `a sunny day`

thanks for your reply. Could you please point out the code implementations where you removed \</s> in log_probs?

**Comment by hijkzzz** [comment]
> > we also removed `</s>` in log_probs, so the log_probs[:, -num_actions] should be `a sunny day`
> 
> thanks for your reply. Could you please point out the code implementations where you removed </s> in log_probs?

in Actor.forward

**Comment by ZiyiLiubird** [comment]
> > > we also removed `</s>` in log_probs, so the log_probs[:, -num_actions] should be `a sunny day`
> > 
> > 
> > thanks for your reply. Could you please point out the code implementations where you removed  in log_probs?
> 
> in Actor.forward

```python
    def forward(
        self,
        sequences: torch.LongTensor,
        num_actions: int = None,
        attention_mask: Optional[torch.Tensor] = None,
        return_output=False,
    ) -> torch.Tensor:
        """Returns action log probs"""
        output = self.model(sequences, attention_mask=attention_mask)

        if return_output:
            return output
        else:
            log_probs = log_probs_from_logits(output["logits"][:, :-1, :], sequences[:, 1:])
            return log_probs[:, -num_actions:]
```

where?

**Comment by hijkzzz** [comment]
Maybe, there is a bug.
Actually, we need the log_probs of the output of `is a sunny day`  here
I will check and fix it

**Comment by hijkzzz** [comment]
> ```python
> log_probs = log_probs_from_logits(output["logits"][:, :-1, :], sequences[:, 1:])
> ```

see log_probs = log_probs_from_logits(output["logits"][:, :-1, :], sequences[:, 1:])

**Comment by ZiyiLiubird** [comment]
> > ```python
> > log_probs = log_probs_from_logits(output["logits"][:, :-1, :], sequences[:, 1:])
> > ```
> 
> see `log_probs = log_probs_from_logits(output["logits"][:, :-1, :], sequences[:, 1:])`

log_probs = log_probs_from_logits(output["logits"][:, :-1, :], sequences[:, 1:]) can's remove /<\s>, it is a token translation that assure token logits corresponding with its index, since the n-th logits predict the n+1 token in sequences.

**Comment by ZiyiLiubird** [comment]
https://github.com/OpenLLMAI/OpenRLHF/blob/01c78c13ee3ac761085d9d1ba1cd830aa83ae4df/openrlhf/models/actor.py#L123
this operation makes sequences[:,-1] is \</s>

**Comment by hijkzzz** [comment]
` log_probs_from_logits(output["logits"][:, :-1, :], sequences[:, 1:])` only extract the log_probs of input  (n-th) token ` a sunny day`.   and the output token is `sunny day </s>`. we only need to ensure the `</s>` not in input token

**Comment by ZiyiLiubird** [comment]
> Fixed here [37aad73](https://github.com/OpenLLMAI/OpenRLHF/commit/37aad73eab32eab3b847cc7be54216bd7435ca1e)

log_probs[:, -num_actions:] should equals to log_probs[:, len(prompts)-1:] (deepseppd_chat and safe-rlhf versiob, include \</s>) or log_probs[:, len(prompts)-1:-1](openrlhf version, exclusion eos_token). there is still a mismatch. maybe num_actions should not be calculated by action_mask.size(1)

**Comment by hijkzzz** [comment]
`log_probs` should not include `</s>` (as input_token),  the output probs of  `</s>`  is meaningless.

**Comment by ZiyiLiubird** [comment]
but why the output probs of \</s> is meaningless? LLM should learn to output \</s> to decide when to stop generation.

**Comment by ZiyiLiubird** [comment]
I got it, thanks.

**Comment by chenhuixi-1995** [comment]
> Maybe, there is a bug. Actually, we need the log_probs of the output of `is a sunny day` here I will check and fix it

why not "a sunny day"?
